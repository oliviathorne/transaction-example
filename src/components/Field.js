// @flow
import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { registerField, updateFieldValue } from './../actions/form';

const Label = styled.label`
  display: block;
  width: 100%;
  margin-top: 2.5rem;
  letter-spacing: -0.4px;
`;

const Input = styled.input`
  border: ${(props) => (props.error === true ? '2px #d6d7df' : '2px #d30a19')};
  border-style: none none solid none;
  width: 100%;
  min-height: 2.4rem;
  font-size: 1.1rem;
  caret-color: #4b3cfa;

  &:focus {
    outline: none;
    border-bottom: 2px solid #4b3cfa;
  }

  &[type='number'] {
    padding-left: 18px;
    box-sizing: border-box;
  }
`;

const CurrencyIcon = styled.span`
  position: absolute;
  display: block;
  left: 5px;
  top: 27.5px;
  font-weight: 500;
  font-size: 19px;
  z-index: 1;
`;

const InputContainer = styled.div`
  position: relative;
`;

const ErrorMessage = styled.div`
  color: #d30a19;
  margin-top: 0.5rem;
  font-size: 11pt;
`;

type Props = {
  dispatchRegisterField: (string, string) => void,
  dispatchUpdateFieldValue: (string, string) => void,
  handleSubmit: (?boolean | ?string) => void,
  validate: (string, string, number) => void,
  isSubmiting: boolean,
  showCurrencySymbol: boolean,
  name: string,
  label: string,
  fieldValue: string,
  type: string,
  totalAvailable: number
};

type State = {
  valid: boolean | string
};

class Field extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = {
      valid: true
    };
  }

  componentWillMount() {
    this.props.dispatchRegisterField(this.props.name, '');
  }

  componentDidUpdate(prevProps) {
    const { isSubmiting, name, fieldValue, validate, totalAvailable } = this.props;

    if (prevProps.isSubmiting !== isSubmiting && isSubmiting) {
      const IS_VALID = validate(name, fieldValue, totalAvailable);
      this.setState({
        valid: IS_VALID
      });

      this.props.handleSubmit(IS_VALID);
    }
  }

  handleChange = (e) => {
    this.props.dispatchUpdateFieldValue(this.props.name, e.target.value);
  };

  handleBlur = () => {
    const { validate, name, fieldValue, totalAvailable } = this.props;
    const VALID = validate(name, fieldValue, totalAvailable);

    this.setState({
      valid: VALID
    });
  };

  render() {
    const { name, label, type, showCurrencySymbol, fieldValue } = this.props;

    return (
      <InputContainer>
        <Label htmlFor={name}>{label}</Label>
        <Input
          name={name}
          type={type}
          value={fieldValue}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          error={this.state.valid}
        />
        {showCurrencySymbol && <CurrencyIcon>Â£</CurrencyIcon>}
        {this.state.valid !== true && <ErrorMessage>{this.state.valid}</ErrorMessage>}
      </InputContainer>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  totalAvailable: state.form.totalAvailable,
  fieldValue: typeof state.form.fields[ownProps.name] !== 'undefined' ? state.form.fields[ownProps.name].value : ''
});

const mapDispatchToProps = (dispatch) => ({
  dispatchRegisterField: (name, initialValue) => dispatch(registerField(name, initialValue)),
  dispatchUpdateFieldValue: (name, value) => dispatch(updateFieldValue(name, value))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Field);
