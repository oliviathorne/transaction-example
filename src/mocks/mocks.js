export const mockState = {
  form: {
    fields: {},
    savedTransactions: [
      {
        name: {
          value: 'Test name'
        },
        email: {
          value: 'email@test.com'
        },
        amount: {
          value: '3000'
        }
      }
    ],
    startingBalance: 18000,
    totalAvailable: 18000,
    totalSent: 0
  }
};
