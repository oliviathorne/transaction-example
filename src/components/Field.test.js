import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mockState } from './../mocks/mocks';
import Field from './Field';

const mockStore = configureStore();
let store;
let wrapper;

beforeEach(() => {
  store = mockStore(mockState);
  wrapper = mount(
    <Provider store={store}>
      <Field />
    </Provider>
  );
});

describe('Field component', () => {
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('outputs the name and type props into input element', () => {
    const props = {
      name: 'test',
      type: 'text'
    };

    const wrapper = mount(
      <Provider store={store}>
        <Field {...props} />
      </Provider>
    );

    expect(wrapper.find('input[name="test"]').length).toBe(1);
    expect(wrapper.find('input[type="text"]').length).toBe(1);
  });

  it('invokes the validation function on blur', () => {
    let validation = jest.fn();

    let wrapper = mount(
      <Provider store={store}>
        <Field validate={validation} />
      </Provider>
    );

    wrapper
      .find('input')
      .props()
      .onBlur();

    expect(validation).toHaveBeenCalledTimes(1);
  });
});
