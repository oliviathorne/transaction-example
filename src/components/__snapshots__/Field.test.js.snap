// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Field)>
    <Field
      dispatchRegisterField={[Function]}
      dispatchUpdateFieldValue={[Function]}
      fieldValue=""
      totalAvailable={18000}
    >
      <styled.div>
        <StyledComponent
          forwardedClass={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": undefined,
              "componentStyle": ComponentStyle {
                "componentId": "sc-bxivhb",
                "isStatic": true,
                "lastClassName": "cdiCYE",
                "rules": Array [
                  "
  position: relative;
",
                ],
              },
              "displayName": "styled.div",
              "render": [Function],
              "styledComponentId": "sc-bxivhb",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="sc-bxivhb cdiCYE"
          >
            <styled.label>
              <StyledComponent
                forwardedClass={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": undefined,
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bdVaJa",
                      "isStatic": true,
                      "lastClassName": "bDvvZv",
                      "rules": Array [
                        "
  display: block;
  width: 100%;
  margin-top: 2.5rem;
  letter-spacing: -0.4px;
",
                      ],
                    },
                    "displayName": "styled.label",
                    "render": [Function],
                    "styledComponentId": "sc-bdVaJa",
                    "target": "label",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <label
                  className="sc-bdVaJa bDvvZv"
                />
              </StyledComponent>
            </styled.label>
            <styled.input
              error={true}
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            >
              <StyledComponent
                error={true}
                forwardedClass={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": undefined,
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": false,
                      "lastClassName": "HnZQh",
                      "rules": Array [
                        "
  border: ",
                        [Function],
                        ";
  border-style: none none solid none;
  width: 100%;
  min-height: 2.4rem;
  font-size: 1.1rem;
  caret-color: #4b3cfa;

  &:focus {
    outline: none;
    border-bottom: 2px solid #4b3cfa;
  }

  &[type='number'] {
    padding-left: 18px;
    box-sizing: border-box;
  }
",
                      ],
                    },
                    "displayName": "styled.input",
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "input",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                onBlur={[Function]}
                onChange={[Function]}
                value=""
              >
                <input
                  className="sc-bwzfXH HnZQh"
                  onBlur={[Function]}
                  onChange={[Function]}
                  value=""
                />
              </StyledComponent>
            </styled.input>
          </div>
        </StyledComponent>
      </styled.div>
    </Field>
  </Connect(Field)>
</Provider>
`;
